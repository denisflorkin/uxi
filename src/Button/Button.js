/* @flow */

import React, { PropTypes } from 'react';
import radium from 'radium';
import Ripples from '../Motion/Ripples';
import ThemeComponent from '../Base/ThemeComponent';
import CluedInButtonStyle from './Button.style';
import { theme } from '../Theme';


type Props = {
  message: String,
};

class Button extends ThemeComponent<Props> {
  isValidType(type) {
    return (type && (type === 'primary' || type === 'secondary'));
  }

  render() {
    const { message, text, type, click, link, disabled, icon, iconPosition, isFullWidth } = this.props;
    let buttonContent;
    let outerStyle = isFullWidth ? { width: '100%' } : {};
    let iconContentBefore;
    let iconContentAfter;
    const textOrMessage = message || text;
    
    const buttonStyles = [
      this.getStyle('button', CluedInButtonStyle.button),
    ];

    if (this.isValidType(type)) {
      buttonStyles.push(this.getSubStyle('button', type));
    }

    if (disabled) {
      buttonStyles.push(CluedInButtonStyle.disabled);
    } else {
      if (this.isValidType(type)) {
        buttonStyles.push(this.getSubStylePseudoElement('button', type, 'hover'));
      } else {
        buttonStyles.push(this.getPseudoElement('button', 'hover'));
      }
    }

    if (icon) {
      const isHover = radium.getState(this.state, 'button', ':hover');
      let hoverIcon = { color: theme['button'].color };

      if (type === 'primary' || type === 'secondary') {
        hoverIcon = { color: '#fff' };
      }

      if (isHover && type !== 'primary' && type !== 'secondary') {
        hoverIcon = { color: theme['button:hover'].color };
      }

      if (isHover && type === 'primary') {
        hoverIcon = { color: theme['button:primary:hover'].color };
      }

      if (isHover && type === 'secondary') {
        hoverIcon = { color: theme['button:secondary:hover'].color };
      }

      if (iconPosition && iconPosition === 'after') {
        iconContentAfter = (
          <div style={{ position: 'absolute', right: '6px', top: '5px' }}>{React.cloneElement(icon, hoverIcon)}</div>
        );
        buttonStyles.push({ paddingRight: '34px' });
      } else {
        iconContentBefore = (
          <div style={{ position: 'absolute', left: '6px', top: '5px' }}>{React.cloneElement(icon, hoverIcon)}</div>
        );
        buttonStyles.push({ paddingLeft: '34px' });
      }
    }

    if (isFullWidth) {
      buttonStyles.push({ width: '100%' });
    }

    if (link) {
      buttonStyles.push({
        textDecoration: 'none',
        position: 'relative',
        verticalAlign: 'top',
        fontWeight: 'normal',
      });

      return (
        <a onClick={click} style={buttonStyles} href={link}>
          {iconContentBefore}
          <span style={CluedInButtonStyle.text}>{textOrMessage}</span>
          {iconContentAfter}
        </a>
      );
    }

    buttonContent = (
      <button key="button" style={buttonStyles} onClick={click}>
        {iconContentBefore}
        <span style={CluedInButtonStyle.text}>{textOrMessage}</span>
        {iconContentAfter}
      </button>
    );

    return (
      <Ripples style={outerStyle}>
        {buttonContent}
      </Ripples>
    );
  }
}

const havingClicks = (props, propName, componentName) => {
  const clickMethod = props.click || props.onClick;
  
  if (!clickMethod) {
    return new Error(`you must pass either one of click, onClick or link to 'uxi/${componentName}'.`);
  }

  if (!{}.toString.call(props.click) === '[object Function]') {
    if (!props.link) {
      return new Error(`you must pass either one of click or link to 'uxi/${componentName}'.`);
    }
  } else {
    return null; // means props's been validated
  }
};

Button.propTypes = {
  message: PropTypes.node,
  text: PropTypes.string,
  type: PropTypes.string,
  onClick: havingClicks,
  click: havingClicks,
  link: (props, propName, componentName) => {
    if (!props.link && !props.click && !props.onClick) {
      return new Error(`you must pass either one of click or link to '${componentName}'.`);
    }
    if ({}.toString.call(props.link) === '[object String]') {
      return null; // means props's been validated
    }
  },
  disabled: PropTypes.bool,
  icon: PropTypes.node,
  iconPosition: PropTypes.string,
  isFullWidth: PropTypes.bool,
  style: PropTypes.object,
};

Button.contextTypes = {
  theme: PropTypes.object.isRequired,
};

export default radium(Button);
