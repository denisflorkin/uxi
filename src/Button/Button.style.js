import { palette } from '../Theme';

export default {
  baseButton: {
    outline: '0 !important',
    // display: 'inline-block',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: 0,
    textAlign: 'center',
    verticalAlign: 'middle',
    touchAction: 'manipulation',
    cursor: 'pointer',
    backgroundImage: 'none',
    borderWidth: '1px',
    borderStyle: 'solid',
    whiteSpace: 'nowrap',
    paddingTop: '6px',
    paddingBottom: '6px',
    paddingLeft: '12px',
    paddingRight: '12px',
    fontSize: '14px',
    lineHeight: '1.42857143',
    borderRadius: '2px',
    userSelect: 'none',
    transition: 'all 0.5s ease-in',
    textTransform: 'uppercase',
    fontFamily: '\'Fira sans\', sans-serif',
  },
  text: {
    textTransform: 'uppercase',
    letterSpacing: '0.0075em',
  },
  disabled: {
    opacity: '0.6',
    transition: 'none',
    outline: '0 !important',
    cursor: 'not-allowed',
  },
  button: { color: 'darkgrey' },
};
