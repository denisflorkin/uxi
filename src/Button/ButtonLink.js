import React from 'react';
import AvatarWithName from '../Img/AvatarWithName';
import styled from 'styled-components';

const getTypeColor = ({ palette }, type) => {
  if (!type) { return palette.accent.main; }
  if (palette.semantic[type]) { return palette.semantic[type]; }
  return '#fff';
};

const ButtonLinkWrapper = styled.a`
  cursor: pointer;
  color: ${({ theme, type }) => getTypeColor(theme, type)};
  display: inline-block;
  &:hover {
    text-decoration: underline;
  }
  svg {
    color: ${({ theme, type }) => getTypeColor(theme, type)};
    fill: ${({ theme, type }) => getTypeColor(theme, type)};
  }
`;

const ButtonLink = ({ onClick, icon, text }) => (
  <ButtonLinkWrapper onClick={(e) => { e.preventDefault(); onClick(e); }}>
    <AvatarWithName imgSize={16} icon={icon} name={text} />
  </ButtonLinkWrapper>
);

export default ButtonLink;
