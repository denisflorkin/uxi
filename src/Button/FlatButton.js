/* @flow */
/* eslint-disable brace-style */
import React, { Component } from 'react';
import styled, { css } from 'styled-components';
import Ripples from '../Motion/Ripples';
import ButtonBaseStyles from './ButtonBaseStyles';

const getTypeColor = ({ palette }, type) => {
  if (palette.semantic[type]) { return palette.semantic[type]; }
  if (type === 'primary') { return palette.accent.main; }
  if (type === 'secondary') { return palette.primary.main; }
  return '#fff';
};

/* eslint-disable indent */
const ButtonBaseMixin = css`
  /* STATIC BASE STYLES: */
  ${ButtonBaseStyles};

  /* ICON POSITION: */
  flex-direction: ${({ iconPosition }) => {
    if (iconPosition && iconPosition === 'after') { return 'row-reverse'; }
    return 'row';
  }};

  /* FULL WIDTH STYLES: */
  ${({ isFullWidth }) => (isFullWidth ? 'width: 100%' : '')};

  /* TYPE STYLES: */
  background-color: transparent;
  border-color: transparent;
  color: ${({ theme, type }) => (type ? getTypeColor(theme, type) : theme.palette.darkGrey)};
  /* this next line overwritte the '.root a' selector rules from uxi NON-StyledComponent theme/css */
  * { color: ${({ theme, type }) => (type ? getTypeColor(theme, type) : theme.palette.darkGrey)}; }
  svg { fill: ${({ theme, type }) => (type ? getTypeColor(theme, type) : theme.palette.darkGrey)};}
  &:hover {
    border-color: transparent;
    color: ${({ theme, type }) => '#ffffff' };
    background-color: ${({ type, theme }) => (type ? getTypeColor(theme, type) : theme.palette.grey)};
    * { color: inherit; }
    svg { fill: ${({ type, theme }) => (type ? '#ffffff' : '#ffffff') }
  }

  /* DISABLED STYLES: (overrides types styles)*/
  cursor: ${({ disabled }) => (disabled ? 'normal' : 'pointer')};
  ${({ disabled, theme: { palette } }) => (disabled ? `background-color: ${palette.lightGrey}; color: ${palette.grey};` : '')}
  ${({ disabled }) => (disabled ? 'border-color: transparent;' : '')};

  &:hover {
    ${({ disabled, theme: { palette } }) => (disabled ? `background-color: ${palette.lightGrey}; color: ${palette.grey}` : '')};
    ${({ disabled }) => (disabled ? 'border-color: transparent;' : '')};
    svg {
      ${({ disabled, theme: { palette } }) => (disabled ? `fill: ${palette.grey}` : '')};
    }
  }
  svg {
    ${({ disabled, theme: { palette } }) => (disabled ? `fill: ${palette.grey}` : '')};
  }

}
`;

const ButtonIconWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: ${({ icon, textOrMessage }) =>
    (icon && textOrMessage ? '0 4px' : '0')};
`;

const ButtonContentWrapper = styled.div`
  margin: 0 4px;
  line-height: 0.7;
`;


/* eslint-enable indent */
const ButtonUI = styled.button`${ButtonBaseMixin};`;
const ButtonLinkUI = styled.a`
  ${ButtonBaseMixin};
  text-decoration: none
`;
const ButtonDivUI = styled.div`${ButtonBaseMixin};`;

// eslint-disable-next-line react/prefer-stateless-function
class Button extends Component {
  render() {
    const {
      message,
      text,
      type: originalType,
      children,
      inert,
      link,
      isFullWidth,
      click,
      onClick,
      disabled,
      icon: iconProp,
      iconPosition,
      style,
      className,
      target,
      // ...restOfProps
    } = this.props;

    const textOrMessage = message || text || children;

    const wasASubmitInitially = originalType === 'submit';

    const type = wasASubmitInitially ? 'primary' : originalType;


    let icon;
    if (React.isValidElement(iconProp)) {
      icon = React.cloneElement(iconProp, { size: '16' });
    }

    const buttonAttr = {
      onClick: click || onClick || null,
      className,
      style,
    };

    const styleProps = {
      isFullWidth,
      disabled,
      type,
      iconPosition,
      icon,
      textOrMessage,
    };

    // which element to render div|a|button
    let TheButtonComponent = null;
    if (inert) { TheButtonComponent = ButtonDivUI; }
    else if (link) { TheButtonComponent = ButtonLinkUI; }
    else { TheButtonComponent = ButtonUI; }

    const theButton = (
      <TheButtonComponent
        {...(link && !disabled ? { href: link } : {})}
        {...(link && !disabled && target ? { target } : {})}
        {...styleProps}
        {...buttonAttr}
      >
        {icon &&
          <ButtonIconWrapper {...styleProps}> {icon} </ButtonIconWrapper>
        }
        {textOrMessage &&
          <ButtonContentWrapper {...styleProps}> {textOrMessage} </ButtonContentWrapper>
        }
      </TheButtonComponent>
    );

    const rippleStyles = isFullWidth ? { width: '100%' } : {};

    return (<Ripples disabled={disabled} style={rippleStyles}>{theButton}</Ripples>);
    // return disabled ? theButton : (<Ripples style={rippleStyles}>{theButton}</Ripples>);
  }
}

export default Button;
