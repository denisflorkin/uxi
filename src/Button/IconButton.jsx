/* @flow */
import React, { Component } from 'react';
import radium from 'radium';
// import Ripples from '../Motion/Ripples';
import Icon from '../Icons';
import getAppropriateIcon from '../Icons/getAppropriateIcon';

class IconButton extends Component {
  render() {
    const {
      onClick,
      link,
      disabled,
      icon,
    } = this.props;

    const Icon = getAppropriateIcon(icon);

    if (link) {
      <a href={link}>
        <Icon {...this.props} />
      </a>;
    }

    return (
      <button style={{ border: 0, padding: 0, verticalAlign: 'middle', cursor: 'pointer', background: 'transparent' }} onClick={onClick}>
        <Icon {...this.props} />
      </button>
    );
  }
}

export default radium(IconButton);
