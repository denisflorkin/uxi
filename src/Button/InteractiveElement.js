/* @flow */
/** when you need to put dynamic handler on a div, use this */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
// import styled from 'styled-components';


// const InteractiveElementUI = styled.button`
//   border: none;
//   display: block;
//   padding: 0;
//   font-size: inherit;
//   background: transparent;
// `;

const styles = {
  border: 'none',
  display: 'block',
  padding: 0,
  margin: 0,
  fontSize: 'inherit',
  background: 'transparent',
};

class InteractiveElement extends PureComponent {
  render() {
    const { value, style, ...restOfProps } = this.props;

    return (
      <button style={{ ...styles, ...style }} {...restOfProps} />
    );
  }
}

InteractiveElement.displayName = 'InteractiveElement';

InteractiveElement.propTypes = {
  innerRef: PropTypes.func,
  ref: PropTypes.func,
  value: PropTypes.any,
  style: PropTypes.object,
};

InteractiveElement.defaultProps = {
  innerRef: null,
  ref: null,
  value: null,
  style: {},
};

export default InteractiveElement;
