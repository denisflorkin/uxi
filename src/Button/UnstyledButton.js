/* @flow */
/* eslint-disable brace-style */
import React, { Component } from 'react';
import styled from 'styled-components';
import Ripples from '../Motion/Ripples';


const UnstyledButtonUI = styled.button`
  border: none;
  padding: 0;
  font-size: inherit;
  background: transparent;
  min-width: 100%;
  min-height: 100%;
`;

const UnstyledButton = ({ children, style, onClick, inline }) => {
  'r';

  if (inline) {
    return (
      <div style={{ minWidth: '100%', minHeight: '100%' }}>
        <UnstyledButtonUI onClick={onClick} style={style}>
          {children}
        </UnstyledButtonUI>
      </div>
    );
  }

  return (
    <div style={{ minWidth: '100%', minHeight: '100%', display: 'flex' }}>
      <UnstyledButtonUI onClick={onClick} style={style}>
        {children}
      </UnstyledButtonUI>
    </div>
  );
};

export default UnstyledButton;
