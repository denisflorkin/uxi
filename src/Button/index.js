import Button from './Button';

export Button from './Button';
export ButtonWithoutRipple from './ButtonWithoutRipple';
export FlatButton from './FlatButton';
export UnstyledButton from './UnstyledButton';
export IconButton from './IconButton';
export ButtonLink from './ButtonLink';

export default Button;
