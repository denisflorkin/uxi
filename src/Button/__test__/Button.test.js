import React from 'react';
import Button from '../Button';
import ThemeProvider from '../../Theme';

it('should render a Button', () => {
  const testValue = 'YO!';
  const wrapper = shallow(
    <ThemeProvider>
      <Button onClick={()=>{}} text={testValue}/>
    </ThemeProvider>
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.props().text).toContain(testValue);
});
