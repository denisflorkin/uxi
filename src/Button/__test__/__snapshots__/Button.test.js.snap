// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Button 1`] = `
<Button
  onClick={[Function]}
  text="YO!"
/>
`;
